app/models/service.rb

```diff
@@ -0,0 +1,35 @@
+class Service < Aggregate
+  MAXIMUM_COMPONENTS_FOR_SERVICE = 2
+  SERVICE_PROVIDERS = /^VSP|SP/.freeze
+
+  has_and_belongs_to_many :components, before_add: %i[maximum_components_for_service
```

Comment:
also, have you considered using `has_many through`? might be easier for the validation logic